### PR TITLE
[PAC] Implement function pointer re-signing

### PR DESCRIPTION
**Pull request by ahatanak**
Re-signing occurs when function type discrimination is enabled and a function pointer is converted to another function pointer type that requires signing using a different discriminator. A function pointer is re-signed using discriminator zero when it's converted to a pointer to a non-function type such as `void*`.

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 1]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
...
+CodeGenFunction::EmitPointerAuthResignCall(llvm::Value *value,
+                                           const CGPointerAuthInfo &curAuth,
+                                           const CGPointerAuthInfo &newAuth) {
+  assert(curAuth && newAuth);
+
+  if (curAuth.getAuthenticationMode() !=
+          PointerAuthenticationMode::SignAndAuth ||
+      newAuth.getAuthenticationMode() !=
+          PointerAuthenticationMode::SignAndAuth) {
+    auto authedValue = EmitPointerAuthAuth(curAuth, value);
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.cpp:3165]: Nit: either

```suggestion
    llvm::Value *authedValue = EmitPointerAuthAuth(curAuth, value);
```
or
```suggestion
    auto *authedValue = EmitPointerAuthAuth(curAuth, value);
```

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 2]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
...
+    return EmitPointerAuthSign(newAuth, authedValue);
+  }
+  // Convert the pointer to intptr_t before signing it.
+  auto origType = value->getType();
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.cpp:3169]: Nit
```suggestion
  auto *origType = value->getType();
```

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 3]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
...
+  value = Builder.CreatePtrToInt(value, IntPtrTy);
+
+  auto curKey = Builder.getInt32(curAuth.getKey());
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.cpp:3172]: Nit
```suggestion
  auto *curKey = Builder.getInt32(curAuth.getKey());
```

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 4]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
...
+  auto newKey = Builder.getInt32(newAuth.getKey());
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.cpp:3173]: Nit
```suggestion
  auto *newKey = Builder.getInt32(newAuth.getKey());
```

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 5]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
...
+    curDiscriminator = Builder.getSize(0);
+
+  llvm::Value *newDiscriminator = newAuth.getDiscriminator();
+  if (!newDiscriminator)
+    newDiscriminator = Builder.getSize(0);
+
+  // call i64 @llvm.ptrauth.resign(i64 %pointer,
+  //                               i32 %curKey, i64 %curDiscriminator,
+  //                               i32 %newKey, i64 %newDiscriminator)
+  auto intrinsic = CGM.getIntrinsic(llvm::Intrinsic::ptrauth_resign);
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.cpp:3186]: Nit
```suggestion
  auto *intrinsic = CGM.getIntrinsic(llvm::Intrinsic::ptrauth_resign);
```

File: clang/test/CodeGen/ptrauth-function-lvalue-cast-disc.c [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  // CHECK: [[LOAD:%.*]] = load ptr, ptr @cptr
+  // CHECK-NEXT: [[CAST4:%.*]] = ptrtoint ptr [[LOAD]] to i64
+  // CHECK-NEXT: [[RESIGN:%.*]] = call i64 @llvm.ptrauth.resign(i64 [[CAST4]], i32 0, i64 0, i32 0, i64 18983)
+  // CHECK-NEXT: [[CAST5:%.*]] = inttoptr i64 [[RESIGN]] to ptr
+  // CHECK-NEXT: call void [[CAST5]]() [ "ptrauth"(i32 0, i64 18983) ]
+}
+
+void *vptr;
+void test5() {
```
Comment (kovdan01) [inline clang/test/CodeGen/ptrauth-function-lvalue-cast-disc.c:41]: Have you missed CHECK-LABEL or is that done intentionally?

File: clang/test/CodeGen/ptrauth-function-lvalue-cast-disc.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// CHECK-LABEL: define i8 @test2
+char test2() {
+  return *(char *)fptr;
+
+  // CHECK: [[LOAD:%.*]] = load ptr, ptr @fptr
+  // CHECK: [[CMP:%.*]] = icmp ne ptr [[LOAD]], null
+
+  // CHECK: [[TOINT:%.*]] = ptrtoint ptr [[LOAD]] to i64
```
Comment (kovdan01) [inline clang/test/CodeGen/ptrauth-function-lvalue-cast-disc.c:25]: Here, you don't check the full function body, and only partially check instructions after `resign.nonnull:` label. The `resign.cont:` label is left unchecked. Was it done intentionally?

File: clang/lib/CodeGen/CodeGenFunction.h [hunk 8]
```diff
@@ -2687,7 +2687,8 @@ class CodeGenFunction : public CodeGenTypeCache {
     if (Alignment.isZero())
       Alignment =
           CGM.getNaturalTypeAlignment(T, BaseInfo, TBAAInfo, ForPointeeType);
-    return Address(Ptr, ConvertTypeForMem(T), Alignment, nullptr,
+    return Address(Ptr, ConvertTypeForMem(T), Alignment,
+                   CGM.getPointerAuthInfoForPointeeType(T), nullptr,
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.h:2691]: Nit
```suggestion
                   CGM.getPointerAuthInfoForPointeeType(T), /*Offset=*/nullptr,
```

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 9]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
...
+
+  auto curKey = Builder.getInt32(curAuth.getKey());
+  auto newKey = Builder.getInt32(newAuth.getKey());
+
+  llvm::Value *curDiscriminator = curAuth.getDiscriminator();
+  if (!curDiscriminator)
+    curDiscriminator = Builder.getSize(0);
+
+  llvm::Value *newDiscriminator = newAuth.getDiscriminator();
+  if (!newDiscriminator)
```
Comment (kovdan01) [inline clang/lib/CodeGen/CodeGenFunction.cpp:3180]: Nit: it looks like that it's a pretty common practice to check returned value of `.getDiscriminator()` against nullptr and then assign discriminator to zero constant of size type if the check succeeds. Probably it's worth adding a corresponding file-local helper or member function.

Feel free to ignore.

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 10]
```diff
@@ -351,3 +434,125 @@ CodeGenModule::getVTablePointerAuthInfo(CodeGenFunction *CGF,
...
+  if (CurInfo == NewInfo && !hasOffset())
+    Val = getBasePointer();
+  else {
+    assert(!Offset && "unexpected non-null offset");
+    Val = CGF.EmitPointerAuthResign(getBasePointer(), QualType(), CurInfo,
+                                    NewInfo, isKnownNonNull());
+  }
+
+  Val = CGF.Builder.CreateBitCast(Val, getType());
+  return Address(Val, getElementType(), getAlignment(), NewInfo, nullptr,
```
Comment (kovdan01) [inline clang/lib/CodeGen/CGPointerAuth.cpp:540]: Nit
```suggestion
  return Address(Val, getElementType(), getAlignment(), NewInfo, /*Offset=*/nullptr,
```

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 11]
```diff
@@ -351,3 +434,125 @@ CodeGenModule::getVTablePointerAuthInfo(CodeGenFunction *CGF,
...
+                   isKnownNonNull());
+
+  assert(ElementType && "Effective type has to be set");
+
+  // If the current and the new ptrauth infos are the same and the offset is
+  // null, just cast the base pointer to the effective type.
+  if (CurInfo == NewInfo && !hasOffset())
+    Val = getBasePointer();
+  else {
+    assert(!Offset && "unexpected non-null offset");
```
Comment (kovdan01) [inline clang/lib/CodeGen/CGPointerAuth.cpp:534]: Nit: It looks like that the assertion can be moved above, before if statement. Currently, if `hasOffset()`, we go to the else branch, and the assertion fails, so there should be no difference in observable behavior if we move the assertion above if.

Feel free to ignore.

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 12]
```diff
@@ -165,6 +166,88 @@ CGPointerAuthInfo CodeGenModule::getPointerAuthInfoForType(QualType T) {
...
+    contBB = createBasicBlock("resign.cont");
+    resignBB = createBasicBlock("resign.nonnull");
+
+    auto isNonNull = Builder.CreateICmpNE(value, null);
+    Builder.CreateCondBr(isNonNull, resignBB, contBB);
+    EmitBlock(resignBB);
+  }
+
+  // Perform the auth/sign/resign operation.
+  if (!newAuthInfo) {
```
Comment (kovdan01) [inline clang/lib/CodeGen/CGPointerAuth.cpp:231]: Nit: you can omit braces in the whole if/else if/else chain. See https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by kovdan01 — COMMENTED** [review]
I've left a bunch of mostly cosmetic comments and minor nits which are non-blocking in terms of merging that. However, I'll take another look at this "with fresh eyes" within a day - I need a bit more time to drill into actual logic.

**Review by kovdan01 — COMMENTED** [review]
Mostly LGTM. There are several new pretty minor comments, but they can be addressed in a follor-up PR.

The thing which is better to be addressed as a part of this PR is deleting three unneeded member functions from `CodeGenFunction`. IMHO there is no reason to ship this changes right now. See https://github.com/llvm/llvm-project/pull/98847#discussion_r1679235930

File: clang/lib/CodeGen/CGExprScalar.cpp [hunk 13]
```diff
@@ -2373,7 +2373,9 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
       DestLV.setTBAAInfo(TBAAAccessInfo::getMayAliasInfo());
       return EmitLoadOfLValue(DestLV, CE->getExprLoc());
     }
-    return Builder.CreateBitCast(Src, DstTy);
+
+    llvm::Value *Result = Builder.CreateBitCast(Src, DstTy);
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/lib/CodeGen/CGExprScalar.cpp:2377]: Yes, a couple of tests including `test_cast_from_opaque` test this path.

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 14]
```diff
@@ -3126,3 +3137,57 @@ CodeGenFunction::EmitPointerAuthAuth(const CGPointerAuthInfo &PointerAuth,
   return EmitPointerAuthCommon(*this, PointerAuth, Pointer,
                                llvm::Intrinsic::ptrauth_auth);
 }
+
+llvm::Value *CodeGenFunction::EmitPointerAuthSign(QualType pointeeType,
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/lib/CodeGen/CodeGenFunction.cpp:3141]: I don't think there is any logic behind this choice. The ptrauth-related functions in CodeGenFunction.cpp should be moved to CGPointerAuth.cpp.

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 15]
```diff
@@ -165,6 +166,88 @@ CGPointerAuthInfo CodeGenModule::getPointerAuthInfoForType(QualType T) {
   return ::getPointerAuthInfoForType(*this, T);
 }
 
+static bool isZeroConstant(llvm::Value *value) {
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/lib/CodeGen/CGPointerAuth.cpp]: I fixed the names of variables and functions added in this PR.

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 16]
```diff
@@ -165,6 +166,88 @@ CGPointerAuthInfo CodeGenModule::getPointerAuthInfoForType(QualType T) {
...
+    return ci->isZero();
+  return false;
+}
+
+static bool equalAuthPolicies(const CGPointerAuthInfo &left,
+                              const CGPointerAuthInfo &right) {
+  if (left.isSigned() != right.isSigned())
+    return false;
+  assert(left.isSigned() && right.isSigned() &&
+         "should only be called with non-null auth policies");
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/lib/CodeGen/CGPointerAuth.cpp]: I moved the assertion to the beginning of the function to make it clear that at least one side has to be signed.

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 17]
```diff
@@ -165,6 +166,88 @@ CGPointerAuthInfo CodeGenModule::getPointerAuthInfoForType(QualType T) {
...
+  } else {
+    assert(value->getType()->isIntegerTy());
+    null = llvm::ConstantInt::get(IntPtrTy, 0);
+  }
+  if (value == null) {
+    return value;
+  }
+
+  // If both schemas sign the same way, we're done.
+  if (equalAuthPolicies(curAuthInfo, newAuthInfo)) {
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/lib/CodeGen/CGPointerAuth.cpp]: I don't think we can do that as the function shouldn't return here when different discriminators are used. Also we are going to add more stuff to `equalAuthPolicies` in the future, so I'll leave it as.

File: clang/test/CodeGen/ptrauth.c [hunk 18]
```diff
@@ -0,0 +1,77 @@
+// RUN: %clang_cc1 -fptrauth-function-pointer-type-discrimination -triple arm64-apple-ios -fptrauth-calls -fptrauth-intrinsics -emit-llvm %s  -o - | FileCheck -check-prefix=CHECK -check-prefix=NOPCH %s
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/test/CodeGen/ptrauth.c:1]: Sorry, I just realized the test cases in `ptrauth.c` are duplicates of the ones added to `CodeGen/ptrauth-function-type-discriminator.c` and `CodeGen/ptrauth-function.c`. I'll remove the file.

File: clang/lib/CodeGen/CGPointerAuth.cpp [hunk 19]
```diff
@@ -351,3 +434,125 @@ CodeGenModule::getVTablePointerAuthInfo(CodeGenFunction *CGF,
...
+  }
+
+  Val = CGF.Builder.CreateBitCast(Val, getType());
+  return Address(Val, getElementType(), getAlignment(), NewInfo, nullptr,
+                 isKnownNonNull());
+}
+
+llvm::Value *LValue::getPointer(CodeGenFunction &CGF) const {
+  assert(isSimple());
+  return emitResignedPointer(getType(), CGF);
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/lib/CodeGen/CGPointerAuth.cpp:571]: Yes, we can check whether there is a noticeable regression in compile time.

File: clang/test/CodeGen/ptrauth-function-lvalue-cast-disc.c [hunk 20]
```diff
@@ -0,0 +1,55 @@
+// RUN: %clang_cc1 %s -triple arm64e-apple-ios13 -fptrauth-calls -fptrauth-intrinsics -emit-llvm -o-  -fptrauth-function-pointer-type-discrimination | FileCheck %s
```
[the remark this answers is not in the document]
Comment (ahatanak) [reply clang/test/CodeGen/ptrauth-function-lvalue-cast-disc.c:1]: Thanks! Please add a line to test linux after this PR is merged.

**Review by kovdan01 — APPROVED** [review]
LGTM, thanks @ahatanak !

**Pull request merged**